### ISSUE TITLE
fatal error: error in backend: Unable to insert indirect branch

### ISSUE THREAD

**Issue by pentelbart** [CTSRD-CHERI/llvm-project#513]
When using cheribuild to build a debug qtwebkit with dynamic linkage, a fatal error occurs when attempting to compile the Javascript Core, specifically when compiling the Low Level Interpreter. This doesn't occur when attempting a RelWithDebInfo build

```
fatal error: error in backend: Unable to insert indirect branch
PLEASE submit a bug report to https://github.com/CTSRD-CHERI/llvm-project/issues and include the crash backtrace, preprocessed source, and associated run script.
If you built clang from source, please provide a reduced test case by running:
        $LLVM_BUILD_DIR/bin/creduce_crash_testcase.py <reproducer>.sh
Stack dump:
0.      Program arguments: /home/psb34/cheri/output/sdk/bin/clang++ --target=riscv64-unknown-freebsd13 --sysroot=/home/psb34/cheri/output/rootfs-riscv64-purecap -DBUILDING_JavaScriptCore -DBUILDING_QT__=1 -DBUILDING_WITH_CMAKE=1 -DHAVE_CONFIG_H=1 -DJavaScriptCore_EXPORTS -DQT_ASCII_CAST_WARNINGS -DQT_CORE_LIB -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_BEFORE=0x050000 -DQT_NO_CAST_TO_ASCII -DQT_NO_DYNAMIC_CAST -DQT_NO_EXCEPTIONS -DQT_USE_QSTRINGBUILDER -ISource/JavaScriptCore -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore -I. -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/.. -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/API -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/ForwardingHeaders -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/assembler -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/b3 -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/b3/air -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/bindings -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/builtins -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/bytecode -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/bytecompiler -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/dfg -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/disassembler -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/disassembler/udis86 -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/ftl -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/heap -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/debugger -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/inspector -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/inspector/agents -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/inspector/augmentable -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/inspector/remote -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/interpreter -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/jit -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/llint -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/parser -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/profiler -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/replay -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/runtime -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/tools -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/wasm -I/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/yarr -IDerivedSources -IDerivedSources/ForwardingHeaders -IDerivedSources/JavaScriptCore -IDerivedSources/JavaScriptCore/inspector -I/home/psb34/cheri/qt5/qtwebkit/Source/WTF -isystem /home/psb34/cheri/output/rootfs-riscv64-purecap/usr/local/riscv64-purecap/include -isystem /home/psb34/cheri/output/rootfs-riscv64-purecap/usr/local/riscv64-purecap/include/QtCore -isystem /home/psb34/cheri/output/rootfs-riscv64-purecap/usr/local/riscv64-purecap/./mkspecs/devices/freebsd-generic-clang -target riscv64-unknown-freebsd13 --sysroot=/home/psb34/cheri/output/rootfs-riscv64-purecap -B/home/psb34/cheri/output/sdk/bin -march=rv64imafdcxcheri -mabi=l64pc128d -mno-relax -ggdb -gline-tables-only -Wno-unused-command-line-argument -Werror=implicit-function-declaration -Werror=format -Werror=incompatible-pointer-types -Wno-error -Wno-error=cheri-bitwise-operations -Wno-error=cheri-capability-misuse -Wno-error=format -Wno-deprecated-copy -Wno-error=cheri-capability-misuse -Werror=cheri-bitwise-operations -Werror=cheri-prototypes -Werror=pass-failed -Werror=undefined-internal -fno-exceptions -fno-strict-aliasing -fno-rtti -std=c++11 -fcolor-diagnostics -Qunused-arguments -g -fPIC -Wall -Wextra -Wcast-align -Wformat-security -Wmissing-format-attribute -Wpointer-arith -Wundef -Wwrite-strings -Wno-parentheses-equality -fPIC -pthread -MD -MT Source/JavaScriptCore/CMakeFiles/JavaScriptCore.dir/llint/LowLevelInterpreter.cpp.o -MF Source/JavaScriptCore/CMakeFiles/JavaScriptCore.dir/llint/LowLevelInterpreter.cpp.o.d -o Source/JavaScriptCore/CMakeFiles/JavaScriptCore.dir/llint/LowLevelInterpreter.cpp.o -c /home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/llint/LowLevelInterpreter.cpp
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module '/home/psb34/cheri/qt5/qtwebkit/Source/JavaScriptCore/llint/LowLevelInterpreter.cpp'.
4.      Running pass 'Branch relaxation pass' on function '@_ZN3JSC5LLInt5CLoop7executeENS_8OpcodeIDEPvPNS_2VMEPNS_14ProtoCallFrameEb'
 #0 0x0000000002ea9bc8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/home/psb34/cheri/output/sdk/bin/clang+++0x2ea9bc8)
 #1 0x0000000002ea7a38 llvm::sys::RunSignalHandlers() (/home/psb34/cheri/output/sdk/bin/clang+++0x2ea7a38)
 #2 0x0000000002e2bb34 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) (/home/psb34/cheri/output/sdk/bin/clang+++0x2e2bb34)
 #3 0x0000000002e2bab2 llvm::CrashRecoveryContext::HandleExit(int) (/home/psb34/cheri/output/sdk/bin/clang+++0x2e2bab2)
 #4 0x0000000002ea48f3 (/home/psb34/cheri/output/sdk/bin/clang+++0x2ea48f3)
 #5 0x0000000001851395 (/home/psb34/cheri/output/sdk/bin/clang+++0x1851395)
 #6 0x0000000002e308a6 llvm::report_fatal_error(llvm::Twine const&, bool) (/home/psb34/cheri/output/sdk/bin/clang+++0x2e308a6)
 #7 0x0000000002e3079b (/home/psb34/cheri/output/sdk/bin/clang+++0x2e3079b)
 #8 0x0000000001d73ddb llvm::RISCVInstrInfo::insertIndirectBranch(llvm::MachineBasicBlock&, llvm::MachineBasicBlock&, llvm::DebugLoc const&, long, llvm::RegScavenger*) const (/home/psb34/cheri/output/sdk/bin/clang+++0x1d73ddb)
 #9 0x00000000022fabe1 (anonymous namespace)::BranchRelaxation::runOnMachineFunction(llvm::MachineFunction&) (/home/psb34/cheri/output/sdk/bin/clang+++0x22fabe1)
#10 0x00000000023eb0fd llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/psb34/cheri/output/sdk/bin/clang+++0x23eb0fd)
#11 0x000000000281fd87 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/psb34/cheri/output/sdk/bin/clang+++0x281fd87)
#12 0x0000000002826368 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/psb34/cheri/output/sdk/bin/clang+++0x2826368)
#13 0x00000000028203f2 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/psb34/cheri/output/sdk/bin/clang+++0x28203f2)
#14 0x00000000030b223a clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) (/home/psb34/cheri/output/sdk/bin/clang+++0x30b223a)
#15 0x000000000385caec clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/psb34/cheri/output/sdk/bin/clang+++0x385caec)
#16 0x0000000003fbf943 clang::ParseAST(clang::Sema&, bool, bool) (/home/psb34/cheri/output/sdk/bin/clang+++0x3fbf943)
#17 0x00000000037b1b73 clang::FrontendAction::Execute() (/home/psb34/cheri/output/sdk/bin/clang+++0x37b1b73)
#18 0x00000000037221d4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/psb34/cheri/output/sdk/bin/clang+++0x37221d4)
#19 0x00000000038568a6 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/psb34/cheri/output/sdk/bin/clang+++0x38568a6)
#20 0x0000000001850be6 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/psb34/cheri/output/sdk/bin/clang+++0x1850be6)
#21 0x000000000184ea20 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) (/home/psb34/cheri/output/sdk/bin/clang+++0x184ea20)
#22 0x000000000360e267 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*, bool*) const::$_1>(long) (/home/psb34/cheri/output/sdk/bin/clang+++0x360e267)
#23 0x0000000002e2ba50 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/psb34/cheri/output/sdk/bin/clang+++0x2e2ba50)
#24 0x000000000360d72a clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*, bool*) const (/home/psb34/cheri/output/sdk/bin/clang+++0x360d72a)
#25 0x00000000035e1c5d clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/home/psb34/cheri/output/sdk/bin/clang+++0x35e1c5d)
#26 0x00000000035e208d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::__1::pair<int, clang::driver::Command const*> >&) const (/home/psb34/cheri/output/sdk/bin/clang+++0x35e208d)
#27 0x00000000035f530c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::__1::pair<int, clang::driver::Command const*> >&) (/home/psb34/cheri/output/sdk/bin/clang+++0x35f530c)
#28 0x000000000184e3e0 main (/home/psb34/cheri/output/sdk/bin/clang+++0x184e3e0)
clang++: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 11.0.0 (https://github.com/CTSRD-CHERI/llvm-project.git ef88c21e79a3b27da20b9e80ad651339439cb988)
Target: riscv64-unknown-freebsd13
Thread model: posix
InstalledDir: /home/psb34/cheri/output/sdk/bin
```
I think I've attached the files the compiler asked me to.

[LowLevelInterpreter-b8a692.sh.txt](https://github.com/CTSRD-CHERI/llvm-project/files/5973506/LowLevelInterpreter-b8a692.sh.txt)

[LowLevelInterpreter-b8a692.cpp.txt](https://github.com/CTSRD-CHERI/llvm-project/files/5973505/LowLevelInterpreter-b8a692.cpp.txt)

**Comment by jrtc27** [comment]
As discussed on Slack, this is believed to be https://reviews.llvm.org/D84833. I already had that cherry-picked to my local branch that adds all the mode-dependent (C)JALR-with-immediate support but dev doesn't have any of that, and fixing it properly in dev only to immediately change it would be a bit annoying.

**Comment by pentelbart** [comment]
Applying the associated patch avoids this error.

**Comment by jrtc27** [comment]
Fixed by #516.
